metricbeat/module/elasticsearch/index/data_xpack.go

```diff
@@ -0,0 +1,332 @@
+			"term_vectors_memory_in_bytes":  c.Int("term_vectors_memory_in_bytes"),
+			"norms_memory_in_bytes":         c.Int("norms_memory_in_bytes"),
+			"points_memory_in_bytes":        c.Int("points_memory_in_bytes"),
+			"doc_values_memory_in_bytes":    c.Int("doc_values_memory_in_bytes"),
+			"index_writer_memory_in_bytes":  c.Int("index_writer_memory_in_bytes"),
+			"version_map_memory_in_bytes":   c.Int("version_map_memory_in_bytes"),
+			"fixed_bit_set_memory_in_bytes": c.Int("fixed_bit_set_memory_in_bytes"),
+		}),
+		"store": c.Dict("store", s.Schema{
+			"size_in_bytes": c.Int("size_in_bytes"),
+		}),
+		"refresh": c.Dict("refresh", s.Schema{
+			"total_time_in_millis": c.Int("total_time_in_millis"),
+		}),
+	}
+
+	cacheStatsSchema = s.Schema{
+		"memory_size_in_bytes": c.Int("memory_size_in_bytes"),
+		"evictions":            c.Int("evictions"),
+		"hit_count":            c.Int("hit_count"),
+		"miss_count":           c.Int("miss_count"),
+	}
+)
+
+func eventsMappingXPack(r mb.ReporterV2, m *MetricSet, info elasticsearch.Info, content []byte) error {
+	err := json.Unmarshal(content, &indicesStruct)
+	if err != nil {
+		return err
+	}
+
+	clusterStateMetrics := []string{"metadata", "routing_table"}
+	clusterState, err := elasticsearch.GetClusterState(m.HTTP, m.HTTP.GetURI(), clusterStateMetrics)
+	if err != nil {
+		return err
+	}
+
+	for name, index := range indicesStruct.Indices {
+		event := mb.Event{}
+		indexStats, err := xpackSchema.Apply(index)
+		if err != nil {
+			continue
+		}
+		indexStats["index"] = name
+
+		err = addClusterStateFields(name, indexStats, clusterState)
+		if err != nil {
+			continue
+		}
+
+		event.RootFields = common.MapStr{
+			"cluster_uuid": info.ClusterID,
+			"timestamp":    common.Time(time.Now()),
+			"interval_ms":  m.Module().Config().Period / time.Millisecond,
+			"type":         "index_stats",
+			"index_stats":  indexStats,
+		}
+
+		event.Index = elastic.MakeXPackMonitoringIndexName(elastic.Elasticsearch)
+		r.Event(event)
+	}
+
+	return nil
+}
+
+func addClusterStateFields(indexName string, indexStats, clusterState common.MapStr) error {
+	indexMetadata, err := getClusterStateMetricForIndex(clusterState, indexName, "metadata")
+	if err != nil {
+		return err
+	}
+
+	indexRoutingTable, err := getClusterStateMetricForIndex(clusterState, indexName, "routing_table")
+	if err != nil {
+		return err
+	}
+
+	// TODO:
+
+	// "index_stats.created",
+	v, err := indexMetadata.GetValue("settings.index.creation_date")
+	if err != nil {
+		return err
+	}
+	c, ok := v.(string)
+	if !ok {
+		return elastic.MakeErrorForMissingField("settings.index.creation_date", elastic.Elasticsearch)
+	}
+
+	created, err := strconv.ParseInt(c, 10, 64)
+	if err != nil {
+		return err
+	}
+	indexStats.Put("created", created)
+
+	// "index_stats.version.created", <--- don't think this is being used in the UI, so can we skip it?
+	// "index_stats.version.upgraded", <--- don't think this is being used in the UI, so can we skip it?
```

Comment:
Who can answer these questions? @chrisronline ?

But even if they are not currently used in the UI, would it be a problem to keep them here?

Reply:
I can't see anywhere those fields are read or used in the UI code and I'd be in favor of option 3. If we aren't showing them, we shouldn't need to index them

Reply:
Just for triple cheeseburger with extra bacon confirmation, @tsullivan WDYT about option 3?

Reply:
Fields like this may be getting into the monitoring data automatically as ES returns it in the API response and monitoring collection stores all the data it can find, unless there are exceptions in the monitoring code. Option 3 seems fine to me

Reply:
> Fields like this may be getting into the monitoring data automatically as ES returns it in the API response and monitoring collection stores all the data it can find, unless there are exceptions in the monitoring code. 

Generally that's true but not in this case. These two fields (and their parent) are being specifically inserted into the monitoring doc: https://github.com/elastic/elasticsearch/blob/43592305a1a1abb318a31216284d5a1b3f0838cf/x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/indices/IndexStatsMonitoringDoc.java#L72-L77

This makes it easy to remove them, which is where the consensus seems to be heading.

Reply:
Created PR to remove these fields from internal Elasticsearch collection: https://github.com/elastic/elasticsearch/pull/33584